### PR TITLE
[Android] Add Android Matter Device App for automation testing

### PR DESCRIPTION
**Pull request by chulspro**
#### Problem
* Need to improve automated testing with Matter device types 
* Difficult to verify Various type of devices

#### Change overview
* Create new matter device app to support automated testing with Android devices (Phone, Tablet and so on)

#### Testing
* 1st build with "./scripts/build/build_examples.py --target android-arm64-virtual-device-app build" command
* Excute Android studio and make project virtual-device-app on /examples
* Install apk file on Android phone such as Galaxy and Pixel
* SmartThings App and Hub can commission virtual-device-app and It is possible to test full automation

### CODE REVIEW

File: examples/st-device-app/android/App/core/common/src/main/res/drawable-xhdpi/smartthings_icon.png
Comment (tcarmelveilleux) [inline examples/st-device-app/android/App/core/common/src/main/res/drawable-xhdpi/smartthings_icon.png:1]: We should not have brand-specific assets in any of the open-source tree files. Examples should not have branding directly in them.

Comment (chulspro) [reply examples/st-device-app/android/App/core/common/src/main/res/drawable-xhdpi/smartthings_icon.png:1]: Thank you for your feedback. Okay, I will replace this icon to common icon.

File: examples/st-device-app/android/java/src/com/samsung/matter/stdeviceapp/DACProvider.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+/*
+ *   Copyright (c) 2021 Project CHIP Authors
```
Comment (electrocucaracha) [inline examples/st-device-app/android/java/src/com/samsung/matter/stdeviceapp/DACProvider.java:2]:
```suggestion
 *   Copyright (c) 2023 Project CHIP Authors
```

File: examples/st-device-app/android/java/src/com/samsung/matter/stdeviceapp/StDeviceApp.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
+/*
+ *   Copyright (c) 2021 Project CHIP Authors
```
Comment (electrocucaracha) [inline examples/st-device-app/android/java/src/com/samsung/matter/stdeviceapp/StDeviceApp.java:2]:
```suggestion
 *   Copyright (c) 2023 Project CHIP Authors
```

File: examples/st-device-app/android/java/src/com/samsung/matter/stdeviceapp/StDeviceApp.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    private void postEvent(int event) {
+        Log.d(TAG, "postEvent : " + event);
+        if (mCallback != null) {
+            mCallback.onEvent(event);
+        }
+    }
+
+    public native void nativeInit();
+
+    // called before Matter server is inited
```
Comment (electrocucaracha) [inline examples/st-device-app/android/java/src/com/samsung/matter/stdeviceapp/StDeviceApp.java:47]:
```suggestion
    // called before Matter server is initiated
```

File: examples/st-device-app/android/java/src/com/samsung/matter/stdeviceapp/StDeviceApp.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    public native void preServerInit();
+
+    // called after Matter server is inited
```
Comment (electrocucaracha) [inline examples/st-device-app/android/java/src/com/samsung/matter/stdeviceapp/StDeviceApp.java:50]:
```suggestion
    // called after Matter server is initiated
```

File: examples/st-device-app/android/java/src/com/samsung/matter/stdeviceapp/StDeviceAppCallback.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
```
Comment (electrocucaracha) [inline examples/st-device-app/android/java/src/com/samsung/matter/stdeviceapp/StDeviceAppCallback.java:3]:
```suggestion
 *    Copyright (c) 2023 Project CHIP Authors
```

File: examples/st-device-app/st-device-common/BUILD.gn [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2020 Project CHIP Authors
```
Comment (electrocucaracha) [inline examples/st-device-app/st-device-common/BUILD.gn:1]:
```suggestion
# Copyright (c) 2023 Project CHIP Authors
```

File: examples/st-device-app/st-device-common/include/CHIPProjectAppConfig.h [hunk 7]
```diff
@@ -0,0 +1,67 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
```
Comment (electrocucaracha) [inline examples/st-device-app/st-device-common/include/CHIPProjectAppConfig.h:3]:
```suggestion
 *    Copyright (c) 2023 Project CHIP Authors
```

**Review by electrocucaracha — COMMENTED** [review]
Most of the headers contain out of dates years. I also noticed that the editor is adding some ending line characters in some files.

File: examples/st-device-app/android/App/app/.gitignore [hunk 8]
```diff
@@ -0,0 +1,2 @@
+/build
```
[the remark this answers is not in the document]
Comment (chulspro) [reply examples/st-device-app/android/App/app/.gitignore:1]: We don't want to include merging of build directory after Android studio building in the same directly. I think It is Okay.

**Review by saurabhst — APPROVED** [review]

**Review by electrocucaracha — APPROVED** [review]
\lgtm

File: examples/st-device-app/android/App/core/data/src/test/java/com/samsung/matter/chipstdeviceapp/core/data/ExampleUnitTest.kt [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+import org.junit.Assert.*
+import org.junit.Test
+
+/**
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+class ExampleUnitTest {
```
Comment (andy31415) [inline examples/st-device-app/android/App/core/data/src/test/java/com/samsung/matter/chipstdeviceapp/core/data/ExampleUnitTest.kt:11]: Are these used/exercised anywhere? The same example seems to be copied and pasted in many places. These files do not seem useful (unless they are required for some build to succeed I guess).

File: examples/st-device-app/android/App/core/common/src/main/java/com/samsung/matter/chipstdeviceapp/core/common/DeepLink.kt [hunk 10]
```diff
@@ -0,0 +1,22 @@
+package com.samsung.matter.chipstdeviceapp.core.common
+
+import androidx.core.net.toUri
+import androidx.navigation.NavDeepLinkRequest
+
+object DeepLink {
+  fun getDeepLinkRequestForQrcodeFragment(setting: String): NavDeepLinkRequest {
+    return NavDeepLinkRequest.Builder.fromUri(
+        "android-app://com.samsung.matter.chipstdeviceapp.feature.qrcode/qrcodeFragment/${setting}"
```
Comment (andy31415) [inline examples/st-device-app/android/App/core/common/src/main/java/com/samsung/matter/chipstdeviceapp/core/common/DeepLink.kt:9]: why are the paths `com.samsung`? If this is to become an SDK supported thing, should this be `com.csa` or `com.matter` or similar?

**Review by andy31415 — COMMENTED** [review]
@chulspro I believe paths should not be `com.samsung`. We are trying to avoid company-specific paths in our examples.

File: examples/st-device-app/android/args.gni [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+chip_device_project_config_include = "<CHIPProjectAppConfig.h>"
+chip_project_config_include = "<CHIPProjectAppConfig.h>"
+chip_system_project_config_include = "<SystemProjectConfig.h>"
+
+chip_project_config_include_dirs =
+    [ "${chip_root}/examples/st-device-app/st-device-common/include" ]
+chip_project_config_include_dirs += [ "${chip_root}/config/standalone" ]
+
+chip_build_libshell = true
```
Comment (andy31415) [inline examples/st-device-app/android/args.gni:27]: is this needed for android?

File: examples/st-device-app/android/java/ClusterChangeAttribute.cpp [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        bool onoff = static_cast<bool>(*value);
+
+        ChipLogProgress(Zcl, "Received on/off command endpoint %d value = %d", static_cast<int>(attributePath.mEndpointId), onoff);
+
+        OnOffManager().PostOnOffChanged(attributePath.mEndpointId, onoff);
+    }
+}
+
+void MatterPostAttributeChangeCallback(const app::ConcreteAttributePath & attributePath, uint8_t type, uint16_t size,
```
Comment (andy31415) [inline examples/st-device-app/android/java/ClusterChangeAttribute.cpp:41]: the path says `android/java` but this is a C++ not even JNI. Is this the right location?

File: examples/st-device-app/st-device-common/include/CHIPProjectAppConfig.h [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+#define CHIP_DEVICE_CONFIG_ENABLE_PAIRING_AUTOSTART 1
+
+// Enable extended discovery, set timeout to 24 hours
+#define CHIP_DEVICE_CONFIG_ENABLE_EXTENDED_DISCOVERY 0
+#define CHIP_DEVICE_CONFIG_EXTENDED_DISCOVERY_TIMEOUT_SECS (24 * 60 * 60)
+
+#define CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONABLE_DEVICE_TYPE 1
+
+#define CHIP_DEVICE_CONFIG_DEVICE_TYPE 0 // 0x0023 = 35 = Video Player
```
Comment (andy31415) [inline examples/st-device-app/st-device-common/include/CHIPProjectAppConfig.h:40]: delete the 35=VideoPlayer comment?

Is 0 the right thing here?

File: scripts/build/builders/android.py [hunk 14]
```diff
@@ -301,20 +308,35 @@ def gradlewBuildExampleAndroid(self):
...
-                    "%s/examples/%s/android/App/" % (self.root,
-                                                     self.app.ExampleName()),
-                    "-PmatterBuildSrcDir=%s" % self.output_dir,
-                    "-PmatterSdkSourceBuild=false",
-                    "-PbuildDir=%s" % self.output_dir,
-                    "assembleDebug",
-                ],
-                title="Building Example " + self.identifier,
-            )
+            if self.app.ExampleName() == "st-device-app":
```
Comment (andy31415) [inline scripts/build/builders/android.py:311]: use `elif` to not increase indent as much.

What is the difference here? why is using `-PbuildDir` not ok for st-app? A comment would help as branching out means we have more things non-similar which is harder to maitain.

**Comment by saurabhst** [comment]
> There is a lot of files with the license header. I think that this should be fixed at least for java files... Also, in the readme I can read that the app is an Android app which allows to commission Matter application. However, it's named "st-device-app", where the "st" is for SmartThings (I guess). Does it mean that this app somehow requires SmartThings ecosystem or it is generic enough to run everywhere without such dependency? If the latter is true, maybe it should be named in a more generic way?

yeah "st" is for SmartThings and agreed it should be renamed in a more generic way and not company specific

File: examples/st-device-app/android/.gn [hunk 15]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2020-2023 Project CHIP Authors
```
[the remark this answers is not in the document]
Comment (chulspro) [reply examples/st-device-app/android/.gn:1]: Thank you for your feedback. This is virtual device app and we installed this to Android phone for automation testing with python scripts. I think we can change name to "virtual-device-app" for more intuitive name and understanding to everyone.

File: examples/st-device-app/android/App/.gitignore [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+/.idea/vcs.xml
+/.idea/caches
+/.idea/libraries
+/.idea/modules.xml
+/.idea/workspace.xml
+/.idea/navEditor.xml
+/.idea/assetWizardSettings.xml
+.idea/**
+.DS_Store
+/build
```
[the remark this answers is not in the document]
Comment (chulspro) [reply examples/st-device-app/android/App/.gitignore:18]: Okay, we can use "out" directory for /build.

File: examples/st-device-app/android/App/core/common/src/main/java/com/samsung/matter/chipstdeviceapp/core/common/DeepLink.kt (hunk 10 above)
Comment (chulspro) [reply examples/st-device-app/android/App/core/common/src/main/java/com/samsung/matter/chipstdeviceapp/core/common/DeepLink.kt:9]: We changed it to "com.matter" same as other examples

Comment (chulspro) [reply examples/st-device-app/android/App/core/common/src/main/java/com/samsung/matter/chipstdeviceapp/core/common/DeepLink.kt:9]: Thank you for your feedback. We changed name space to "com.matter" same as other examples

File: examples/st-device-app/android/App/core/data/src/test/java/com/samsung/matter/chipstdeviceapp/core/data/ExampleUnitTest.kt (hunk 9 above)
Comment (chulspro) [reply examples/st-device-app/android/App/core/data/src/test/java/com/samsung/matter/chipstdeviceapp/core/data/ExampleUnitTest.kt:11]: This is our baseline commit. We will add follow up commit from our implementation. We will use this for Next Unit Test related patch.

File: examples/st-device-app/android/java/ClusterChangeAttribute.cpp (hunk 12 above)
Comment (chulspro) [reply examples/st-device-app/android/java/ClusterChangeAttribute.cpp:41]: This is JNI related code, and other projects also used same path. we just recommended other examples.

File: examples/st-device-app/android/args.gni (hunk 11 above)
Comment (chulspro) [reply examples/st-device-app/android/args.gni:27]: Let me check and get back to you.

File: examples/st-device-app/st-device-common/include/CHIPProjectAppConfig.h (hunk 13 above)
Comment (chulspro) [reply examples/st-device-app/st-device-common/include/CHIPProjectAppConfig.h:40]: We will remove this in the next patch.

File: scripts/build/builders/android.py (hunk 14 above)
Comment (chulspro) [reply scripts/build/builders/android.py:311]: We need to add "-PbuildDir" and Also we can use "elif"

**Comment by saurabhst** [comment]
@andy31415 could you please give another pass on this PR considering the review comments have been addressed? We want to merge this PR soon else we need to rebase / resolve conflict multiple times. Thanks

**Comment by woody-apple** [comment]
@woody-apple to resolve

**Comment by woody-apple** [comment]
@andy31415 to take another look.

**Comment by woody-apple** [comment]
@woody-apple to look at this failing job.

**Review by andy31415 — APPROVED** [review]

**Pull request merged**